Add manual SDPmessage definition

notbdu opened the pull request:
- Removes auto generated `SDPMessage`
- Adds a manual `SDPMessage` definition

notbdu commented on gstreamer-sdp/src/s_d_p_key.rs:
```diff
@@ -0,0 +1,11 @@
+
+use ffi;
+
+pub struct SDPKey(ffi::GstSDPKey);
```
I wasn't sure if just wrapping the ffi definition is fine here or not. There were no methods defined in the  -sys repo so I figured this was fine?

Also, I tried to auto gen these definition (and others) but I ran into the same error:
```
thread 'main' panicked at 'Missing memory management functions for GstSdp.SDPKey', src/codegen/record.rs:61:9
```
I'm guessing that's part of why these definitions had to be manually written?

sdroege replied:
Just wrapping the FFI definition is fine but you'll have to implement memory management correctly and also implement the relevant traits from glib-rs. I didn't check if you did that here yet, just saying :) I'll take a closer look later.

For the memory management error, yes that's one of the reasons why these have to be manually implemented.

notbdu commented on Gir_GstSdp.toml:
```diff
@@ -16,6 +16,7 @@ external_libraries = [
 ]
 
 generate = [
+    "GstSdp.MIKEYMessage",
```
Not sure if it was okay to auto gen this definition.

sdroege commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 **/*.rs.bk
 Cargo.lock
+# vim swap files
+*.swp
```
Please add this in a separate commit, or even as a separate PR against master :)

sdroege commented on gstreamer-sdp/src/s_d_p_zone.rs:
```diff
@@ -0,0 +1,27 @@
+pub struct SDPZone(ffi::GstSDPZone);
+
+impl SDPZone {
+    pub fn set(&mut self, adj_time: &str, typed_time: &str) -> SDPResult {
```
This is part of the reason why this (and others can't be autogenerated). set/clear are actually constructor and destructor here, and the `SDPResult` return value should become some kind of `Result<_,_>` (I don't think `clear` can even fail). So there should be a `fn new(adj_time, typed_time) -> Result<Self, _>`, and `clear` just is called in `Drop::drop`.

Also there should be accessors for the fields of these structs

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -8,4 +39,389 @@ impl SDPMessage {
+		}
+	}
+
+	pub fn init(&mut self) -> SDPResult {
```
This one here is actually the constructor

sdroege replied:
That is, you call `glib_ffi::g_malloc(mem::size_of::<GstSDPMessage>())`, then `init` and then return that

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -8,4 +39,389 @@ impl SDPMessage {
+        }
+    }
+
+    pub fn parse_uri(uri: &str, msg: &mut SDPMessage) -> SDPResult {
```
This and `parse_buffer` are also constructors in disguise

sdroege commented:
Shortly looked over it, thanks a lot :) This goes in the wrong direction but there's still more to do

notbdu commented:
Thanks for the feedback @sdroege! I'm going to hack on it again later tonight for me (UTC -5:00).

notbdu commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,440 @@
+        }
+    }
+
+	pub fn init() -> SDPMessage {
```
So after I changed this to a constructor, I noticed that this is doing essentially the same thing as `new()`.

sdroege replied:
Yeah, init() is unneeded in this case

notbdu commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,440 @@
+    pub fn parse_uri(uri: &str, msg: &mut SDPMessage) -> Result<Self, SDPResult> {
+        assert_initialized_main_thread!();
+		unsafe {
+            let mut msg = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMessage>());
```
So in some places I saw usage of `std::mem::uninitialized` so I wasn't sure if I'm supposed to use that or glib's `g_malloc` method. Although, my operating assumption was that the compiler would infer the memory size to allocate?

sdroege replied:
malloc is needed if it should be heap allocated and there is a Drop impl that frees the data (in all cases when you use the glib_wrapper macro), otherwise you can simply stack allocate it

notbdu commented on gstreamer-sdp/src/s_d_p_time.rs:
```diff
@@ -0,0 +1,54 @@
+        CStr::from_ptr(self.0.stop).to_str().unwrap()
+    }
+
+    // TODO: How to convert glib::GArray into a string slice?
```
I wasn't sure how to convert `glib::GArray` into a string slice. It seems like it _might_ be possible to use `std::slice::from_raw_parts`?

sdroege replied:
Is it an actual string or what is it?
If it is a string, you can use CStr on the data pointer of the GArray.

notbdu replied:
It seems to be an array of `GChars`

sdroege replied:
It's an array of arbitrary data, the question is what GstSDPMessage is exactly putting in there. Which seem to be an array of strings, not a single string. So you could return a `Vec<&str>` here

notbdu replied:
How would I convert this https://docs.rs/glib-sys/0.5.0/glib_sys/struct.GArray.html into a `Vec<&str>`? It have the length there but I don't think I can do something like pointer arithmetic to iterate over all the data elements.

sdroege replied:
```rust
let arr = (*vec).data as *const *const c_char;
let vec = Vec::with_capacity((*arr).len as usize);
for i in 0..len {
    vec.push(CStr::new(arr.offset(i)));
}
```

or something like that

notbdu replied:
Ahhh didn't know about `offset`, thanks @sdroege !

notbdu commented:
Btw, I also wasn't sure what you meant by the following:
> Just wrapping the FFI definition is fine but you'll have to implement memory management correctly and also implement the relevant traits from glib-rs.

Mmm what memory management and traits from `glib-rs` are required by structs like `SDPOrigin` and `SDPKey`. It's hard to tell what additional traits/methods are required since both types belong to an `SDPMessage`.

sdroege commented:
> Mmm what memory management and traits from glib-rs are required by structs like SDPOrigin and SDPKey. It's hard to tell what additional traits/methods are required since both types belong to an SDPMessage.

I'll check when going through the commits, maybe nothing else is needed here after all.

notbdu commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,440 @@
+        }
+    }
+
+    pub fn insert_time(&mut self, idx: i32, time: &mut SDPTime) -> SDPResult {
```
Ah! That would be a by product of blindly copying the auto generated code w/o actually thinking about it haha 😓. My bad, yea it doesn't make sense to make it mutable when I'm not actually manipulating it...

notbdu commented:
Thanks for the comments!

notbdu commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,440 @@
+        }
+    }
+
+    pub fn add_bandwidth(&mut self, bwtype: &str, bandwidth: u32) -> SDPResult {
```
Should I add a helper function somewhere to do that? Where do you think it should live?

sdroege replied:
It would just mean to do manual implementations of these functions and then doing `if res == ffi::GST_SDP_RESULT_OK { Ok(whatever) } else { Err(()) }`

notbdu commented on gstreamer-sdp/src/auto/s_d_p_media.rs:
```diff
@@ -176,10 +178,14 @@ impl SDPMedia {
-    //    unsafe { TODO: call ffi::gst_sdp_media_parse_keymgmt() }
-    //}
+    #[cfg(any(feature = "v1_8_1", feature = "dox"))]
+    pub fn parse_keymgmt(&self) -> (SDPResult, MIKEYMessage) {
```
I'm going to move this out of the auto gen'ed files then.

notbdu commented on gstreamer-sdp/src/auto/m_i_k_e_y_message.rs:
```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    //pub fn new_from_bytes(bytes: /*Ignored*/&glib::Bytes, info: /*Ignored*/&mut MIKEYDecryptInfo, error: /*Ignored*/Option<Error>) -> MIKEYMessage {
```
Ah, so there aren't any memory management functions for `MIKEYDecryptInfo`, `MIKEYEncryptInfo` or `MIKEYMapSRTP`.

notbdu replied:
Also, both decrypt/encrypt info are void pointers. Not sure what we would implement for those struct definitions.

sdroege replied:
As structs around the void pointers, `struct MIKEYDecryptInfo(*const c_void)` or similar

notbdu replied:
To avoid changes to `MIKEYMessage` getting wiped, I'm going to move it out of the auto-gen'ed files.

sdroege replied:
You can keep the autogenerated file and just put the manual implementations of functions into a new file into a second `impl MIKEYMessage` block

notbdu commented:
Btw, I can't compile it locally via `cargo build`. Getting the following error:
```
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/gtk-rs/glib`
    Updating git repository `https://github.com/gtk-rs/sys`
    Updating git repository `https://github.com/sdroege/gstreamer-sys`
error: failed to load source for a dependency on `gstreamer-sys`

Caused by:
  Unable to update https://github.com/sdroege/gstreamer-sys?branch=1.14
```

notbdu commented:
So I realized that the mutable arguments in some of the `insert_*` functions were due to the ffi function requiring mutable arguments so I changed them back.

(e.g. `insert_attribute(&mut self, idx: i32, attr: &mut SDPAttribute)`)

notbdu commented on gstreamer-sdp/src/m_i_k_e_y_message.rs:
```diff
@@ -0,0 +1,206 @@
+    pub fn new_from_bytes(bytes: &mut glib::Bytes, info: &mut MIKEYDecryptInfo, error: Option<Error>) -> MIKEYMessage {
+        unsafe {
+            let mut err = match error {
+                Some(err) => glib_ffi::g_error_new_literal(
```
Wasn't sure how to convert a rust `Error` to a `glib::GError` so I ended up doing this... 😓

sdroege replied:
This should actually be `pub fn new_from_bytes(bytes: &mut glib::Bytes, info: &mut MIKEYDecryptInfo) -> Result<MIKEYMessage, Error>`, like in other functions with `glib::Error`.

See e.g. here https://github.com/sdroege/gstreamer-rs/blob/b6a80b59fa0c32f210267de9fcd1d3adf69f1d9e/gstreamer/src/auto/functions.rs#L155-L157

sdroege commented:
> Btw, I can't compile it locally via cargo build. Getting the following error

Rebase your branch against master, not the 1.14 or webrtc branch. Then it should be fine.

sdroege commented:
> So I realized that the mutable arguments in some of the insert_* functions were due to the ffi function requiring mutable arguments so I changed them back.

They don't just require mutable arguments, they take ownership of their arguments. So instead of a reference they should take the arguments by value. This seems to be the case for all insert and replace functions.

sdroege commented:
I'll go through the other changes later today, thanks :)

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_decrypt_info.rs:
```diff
@@ -0,0 +1,11 @@
+use std::mem;
+use std::ptr;
+
+pub struct MIKEYDecryptInfo(ffi::GstMIKEYDecryptInfo);
```
This probably should get some field accessors at least, maybe a constructor. Not sure how it is actually used :)

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_encrypt_info.rs:
```diff
@@ -0,0 +1,11 @@
+use std::mem;
+use std::ptr;
+
+pub struct MIKEYEncryptInfo(ffi::GstMIKEYEncryptInfo);
```
Same

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_map_s_r_t_p.rs:
```diff
@@ -0,0 +1,9 @@
+use glib::translate::*;
+use glib_ffi;
+use std::mem;
+use std::ptr;
```
This file is basically empty

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_message.rs:
```diff
@@ -53,9 +55,11 @@ impl MIKEYMessage {
-    //pub fn add_payload(&mut self, payload: /*Ignored*/&mut MIKEYPayload) -> bool {
-    //    unsafe { TODO: call ffi::gst_mikey_message_add_payload() }
-    //}
+    pub fn add_payload(&mut self, payload: &mut MIKEYPayload) -> bool {
```
Why the mutable reference?

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_payload.rs:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    pub fn kemac_add_sub(&mut self, newpay: &mut MIKEYPayload) -> bool {
```
Why mutable reference?

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_payload.rs:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    //pub fn t_set(&mut self, type_: MIKEYTSType, ts_value: &[u8]) -> bool {
```
Why is this not generated?

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_payload.rs:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    //pub fn sp_get_param(&self, idx: u32) -> /*Ignored*/Option<MIKEYPayloadSPParam> {
```
This looks useful to have :)

sdroege replied:
Also for a future PR

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -223,15 +215,15 @@ impl SDPMessage {
     }
 
-    pub fn insert_attribute(&mut self, idx: i32, attr: &mut SDPAttribute) -> SDPResult {
+    pub fn insert_attribute(&mut self, idx: i32, attr: &SDPAttribute) -> SDPResult {
```
This is taking ownership of the argument from what I can see, so should be changed to do that. After passing to native code, no drop should be called so it should probably be just forgotten with `mem::forget`

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_message.rs:
```diff
@@ -1,20 +1,22 @@
+use m_i_k_e_y_map_s_r_t_p::MIKEYMapSRTP;
 
 glib_wrapper! {
     pub struct MIKEYMessage(Boxed<ffi::GstMIKEYMessage>);
```
Instead of moving the whole file, you can keep the autogenerated one and only have the additional manual things here

notbdu commented:
Hmm so It looks like I'll have to manually edit all of the `auto` files... 

I wanted to move the `auto` files out after manually editing them so they didn't include the `automatically generated...` comment in the head of the file and remove them from the `.toml`. Since all manual edits would get wiped if someone were to run gir again.

I can move them back if you think that makes sense tho!

sdroege commented:
Don't edit the automatically generated files. You can keep them as-is for the parts that are correct and just add an additional file with the manual code. See e.g. https://github.com/sdroege/gstreamer-rs/blob/master/gstreamer/src/auto/bus.rs and https://github.com/sdroege/gstreamer-rs/blob/master/gstreamer/src/bus.rs#L72

notbdu commented:
Oooohh okay! Sounds good, I'll do that then.

notbdu commented:
What about for methods like the following: 
```
pub fn insert_payload(&mut self, idx: u32, payload: &mut MIKEYPayload) -> bool {
    unsafe {
        from_glib(ffi::gst_mikey_message_insert_payload(self.to_glib_none_mut().0, idx, payload.to_glib_full()))
    }
}
```
This is part of `src/auto/m_i_k_e_y_message.rs`, but you mentioned that we want the signature to be `insert_payload(&mut self, idx: u32, payload: MIKEYPayload)` because we are consuming the payload variable and passing full ownership to gstreamer. Would I... comment out the auto-gen'ed function and write my own outside of the `auto` file?

notbdu commented:
Implementing the commented out functions sounds good to me tho! 👍

sdroege commented:
You can mark specific functions as ignored in the toml, see e.g. https://github.com/sdroege/gstreamer-rs/blob/master/Gir_Gst.toml#L132 . And then implement those correctly in the manual file.

notbdu commented:
Okay, I addressed your comments and also squashed my commits. This _should_ be close? 

I wasn't able to rebase off of master due to some nasty merge conflicts which would cause issues merging into your webrtc WIP branch as well.

notbdu commented:
I didn't put any of the commented out functions in the auto-gen'ed files into the `ignore` section of the gir toml. I just went ahead and manually implemented them outside of the `auto` dir.

sdroege commented:
> I didn't put any of the commented out functions in the auto-gen'ed files into the ignore section of the gir toml. I just went ahead and manually implemented them outside of the auto dir.

That's no problem. Just for the ones that are not commented out but need to be manually implemented you must ignore them explicitly.

sdroege commented:
I've rebased the `webrtc-sdp` branch against master, you should be able to rebase yours against that now and then it should be possible to build

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_message.rs:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    pub fn base64_encode(&mut self) -> Option<String> {
```
Does not have to be a mutable reference

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_message.rs:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    //pub fn get_cs_srtp(&self, idx: u32) -> /*Ignored*/Option<MIKEYMapSRTP> {
```
There are still some ignored functions but let's worry about those once the basics are all merged

sdroege commented on gstreamer-sdp/src/auto/m_i_k_e_y_message.rs:
```diff
@@ -0,0 +1,166 @@
+    //    unsafe { TODO: call ffi::gst_mikey_message_replace_cs_srtp() }
+    //}
+
+    pub fn set_info(&mut self, version: u8, type_: MIKEYType, V: bool, prf_func: MIKEYPRFFunc, CSB_id: u32, map_type: MIKEYMapType) -> bool {
```
There are all kinds of setters in this type but no getters, I assume some field accessors would be needed. But that's also for a future PR

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_decrypt_info.rs:
```diff
@@ -0,0 +1,23 @@
+
+impl MIKEYDecryptInfo {
+    pub fn new() -> MIKEYDecryptInfo {
+        MIKEYDecryptInfo(ffi::GstMIKEYDecryptInfo(ptr::null_mut() as *mut c_void))
```
This and the other constructors that use NULL don't seem like a good idea. Just don't have a constructor, these would have to be created differently somehow... and there does not even seem to be a definition inside GStreamer for these structs, and they are not even used anywhere internally. You can safely make them in function arguments optional (`Option`, can also be configured in the toml).

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_message.rs:
```diff
@@ -0,0 +1,98 @@
+use m_i_k_e_y_map_s_r_t_p::MIKEYMapSRTP;
+
+impl MIKEYMessage {
+    pub fn new_from_bytes(bytes: glib::Bytes, info: MIKEYDecryptInfo) -> Result<MIKEYMessage, Error> {
```
Can take the bytes by reference. The function does not take ownership of it

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_payload.rs:
```diff
@@ -0,0 +1,35 @@
+        mem::forget(newpay);
+    }
+
+    //pub fn sp_get_param(&self, idx: u32) -> /*Ignored*/Option<MIKEYPayloadSPParam> {
```
Why commented out functions in the manual file?

sdroege commented on gstreamer-sdp/src/s_d_p_attribute.rs:
```diff
@@ -0,0 +1,44 @@
+            let mut attr = mem::uninitialized();
+            let result = from_glib(ffi::gst_sdp_attribute_set(&mut attr, key.to_glib_none().0, value.to_glib_none().0));
+            match result {
+                SDPResult::Ok => Ok(SDPAttribute(attr)),
```
As `SDPResult` is an enum with just `OK` and `INVAL`, you can make the return values of all these functions simply `Result<_, ()>`. Less noise, and we don't have to generate the `SDPResult` type at all then (only in the -sys bindings but there we have it already)

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+}
+
+impl SDPMedia {
+    pub fn add_attribute<'a, P: Into<Option<&'a str>>>(&mut self, key: &str, value: P) -> Result<_, ()> {
```
You need to provide a type for the successful case. This should be `Result<(), ()>` I guess

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn bandwidths_len(&self) -> u32 {
```
In a future PR it would make sense to get rid of this and the indexed getter (also for other fields), and return a FixedSizeIterator or a Vec

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn get_attribute(&self, idx: u32) -> Option<SDPAttribute> {
```
This does not give you ownership of the return value but borrows it to you. It should return a `Option<&SDPAttribute>`. Same for all the other getters

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn get_attribute_val(&self, key: &str) -> Option<String> {
```
`Option<&str>`

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn get_caps_from_media(&self, pt: i32) -> Option<gst::Caps> {
```
Only exception: the caps are indeed owned

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn init(&mut self) -> Result<_, ()> {
```
This can go away, you have a constructor

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn insert_attribute(&mut self, idx: i32, attr: &mut SDPAttribute) -> Result<_, ()> {
```
This takes ownership of the argument, not just a mutable reference. And you should `mem::forget` it in the end

sdroege replied:
Same for all the other insert/replace, except for the ones taking strings

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -0,0 +1,448 @@
+        }
+    }
+
+    pub fn uninit(&mut self) -> Result<_, ()> {
```
Should go away, that's basically the drop impl

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,617 @@
             let msg = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMessage>());
-            let _ = ffi::gst_sdp_message_new(msg);
+            ffi::gst_sdp_message_new(msg);
             from_glib_full(msg)
```
This function is actually wrong
```rust
let mut msg = mem::uninitialized();
ffi::gst_sdp_message_new(&mut msg);
from_glib_full(msg)
```

What you implemented here would be correct for calling `gst_sdp_message_init`, if you called `g_malloc0`

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,617 @@
+        }
+    }
+
+    pub fn emails_len(&self) -> u32 {
```
Also these should become a FixedSizeIterator or at least Vec in a future PR

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,617 @@
+        }
+    }
+
+    pub fn get_attribute(&self, idx: u32) -> Option<SDPAttribute> {
```
Borrowing for all getters. You can do this by marking the struct as `#[repr(C)]` and then doing `&*(the_ptr as *mut SDPAttribute)`, i.e. casting from the FFI type to your wrapper struct

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -1,11 +1,617 @@
+        }
+    }
+
+    pub fn parse_uri(uri: &str, msg: &mut SDPMessage) -> Result<Self, SDPResult> {
```
Should not take a `msg` parameter anymore

notbdu commented:
Thanks for the feedback!

notbdu commented:
Just a heads up, `cargo build` is still failing even after the rebase.

```
bdu @ Bos-MacBook-Pro.local gstreamer-sdp (bdu/webrtcsdp) $ cargo build --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/gtk-rs/glib`
    Updating git repository `https://github.com/gtk-rs/sys`
    Updating git repository `https://github.com/sdroege/gstreamer-sys`
    Updating git repository `https://github.com/gtk-rs/gio`
    Updating git repository `https://github.com/sdroege/gstreamer-sys`
error: failed to load source for a dependency on `gstreamer-sys`

Caused by:
  Unable to update https://github.com/sdroege/gstreamer-sys?branch=1.14

Caused by:
  failed to find branch `1.14`

Caused by:
  cannot locate local branch '1.14'; class=Reference (4); code=NotFound (-3)
```

sdroege commented:
Even after cargo update? There should be no reference to the 1.14 branch in any Cargo.toml anymore.

I'll check later

notbdu commented:
Yea.. :/, when I run `cargo update` I see this:
```
bdu @ Bos-MacBook-Pro.local gstreamer-sdp (bdu/webrtcsdp) $ cargo update --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/gtk-rs/glib`
    Updating git repository `https://github.com/gtk-rs/sys`
    Updating git repository `https://github.com/sdroege/gstreamer-sys`
    Updating git repository `https://github.com/gtk-rs/gio`
    Updating git repository `https://github.com/sdroege/gstreamer-sys`
error: failed to load source for a dependency on `gstreamer-sys`

Caused by:
  Unable to update https://github.com/sdroege/gstreamer-sys?branch=1.14

Caused by:
  failed to find branch `1.14`

Caused by:
  cannot locate local branch '1.14'; class=Reference (4); code=NotFound (-3)
```

sdroege commented:
In `gstreamer-webrtc/Cargo.toml`, change that from `branch = "1.14"` to master or complete change the branch parameter.

notbdu commented:
Oh sweet, I can see the > 1 million errors I have now haha. Will update PR w/ fixes when I'm finished.

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_message.rs:
```diff
@@ -25,17 +25,17 @@ use m_i_k_e_y_encrypt_info::MIKEYEncryptInfo;
-            let ret = ffi::gst_mikey_message_new_from_bytes(bytes.to_glib_full(), info.to_glib_full(), &mut error);
+            let ret = ffi::gst_mikey_message_new_from_bytes(bytes.to_glib_none(), info.to_glib_full(), &mut error);
             mem::forget(bytes);
             mem::forget(info);
```
This function does not take ownership of the bytes, same for the other one below with data.

The `info` is also not taken ownership of

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_payload.rs:
```diff
@@ -25,11 +26,21 @@ impl MIKEYPayload {
-    //pub fn sp_get_param(&self, idx: u32) -> /*Ignored*/Option<MIKEYPayloadSPParam> {
-    //    unsafe { TODO: call ffi::gst_mikey_payload_sp_get_param() }
-    //}
+    pub fn kemac_get_sub(&self, idx: u32) -> Option<MIKEYPayload> {
```
This is missing a `&`

sdroege replied:
Still missing, this can't really compile as-is, can it?

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -22,6 +22,7 @@ use s_d_p_connection::SDPConnection;
 use s_d_p_key::SDPKey;
 
 glib_wrapper! {
+    #[repr(C)]
```
This does not need `#[repr(C)]`. For functions to return a reference to an `SDPMedia`, you would use `from_glib_borrow()`

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -31,47 +32,56 @@ glib_wrapper! {
 
 impl SDPMedia {
-    pub fn add_attribute<'a, P: Into<Option<&'a str>>>(&mut self, key: &str, value: P) -> Result<_, ()> {
+    pub fn new() -> (SDPResult, SDPMedia) {
```
`SDPResult` not really needed here, this should be `Result<Self, ()>`

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -118,27 +128,27 @@ impl SDPMedia {
-    pub fn get_attribute_val_n(&self, key: &str, nth: u32) -> Option<String> {
+    pub fn get_attribute_val_n(&self, key: &str, nth: u32) -> Option<&str> {
         unsafe {
             from_glib_none(ffi::gst_sdp_media_get_attribute_val_n(self.to_glib_none().0, key.to_glib_none().0, nth))
```
For getting `&str` references you need to use `CStr` directly

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -118,27 +128,27 @@ impl SDPMedia {
+    pub fn get_bandwidth(&self, idx: u32) -> Option<&SDPBandwidth> {
         unsafe {
-            from_glib_none(ffi::gst_sdp_media_get_bandwidth(self.to_glib_none().0, idx))
+            &*(from_glib_none(ffi::gst_sdp_media_get_bandwidth(self.to_glib_none().0, idx)) as *mut SDPBandwidth)
```
For all the other references, don't use `from_glib_none` either. This one here for example would be
```Rust
let ptr = ffi::gst_sdp_media_get_bandwidth(self.to_glib_none().0, idx);
if ptr.is_null() {
    None
} else {
    Some(&*(ptr as *mut SDPBandwidth))
}
```

(or maybe `as *const ...`)

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -190,249 +200,225 @@ impl SDPMedia {
+    pub fn insert_attribute(&mut self, idx: i32, attr: SDPAttribute) -> Result<(), ()> {
         let result = unsafe {
-            from_glib(ffi::gst_sdp_media_insert_attribute(self.to_glib_none_mut().0, idx, attr.to_glib_none_mut().0))
+            from_glib(ffi::gst_sdp_media_insert_attribute(self.to_glib_none_mut().0, idx, attr.to_glib_full()))
```
Not `to_glib_full` for all these as it would create a copy. Instead `&attr.0`

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -190,249 +200,225 @@ impl SDPMedia {
+    pub fn replace_bandwidth(&mut self, idx: u32, bw: SDPBandwidth) -> Result<(), ()> {
         let result = unsafe {
-            from_glib(ffi::gst_sdp_media_replace_bandwidth(self.to_glib_none_mut().0, idx, bw.to_glib_none_mut().0))
+            from_glib(ffi::gst_sdp_media_replace_bandwidth(self.to_glib_none_mut().0, idx, bw.to_glib_full()))
```
Also for all these types you did not implement the GLib translation traits so this won't work. But you don't need the traits here, just some pointer casting for these types

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -35,85 +35,84 @@ impl SDPMessage {
-            ffi::gst_sdp_message_new(msg);
-            from_glib_full(msg)
+            let msg = mem::uninitialized();
+            from_glib_full(ffi::gst_sdp_message_new(&mut msg))
```
This returns an `SDPResult`, not the actual message. Your message would be `msg`

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -165,167 +164,167 @@ impl SDPMessage {
+    pub fn get_attribute(&self, idx: u32) -> Option<&SDPAttribute> {
         unsafe {
-            from_glib_none(ffi::gst_sdp_message_get_attribute(self.to_glib_none().0, idx))
+            &*(from_glib_none(ffi::gst_sdp_message_get_attribute(self.to_glib_none().0, idx)) as *mut SDPAttribute)
```
Just like above, store the pointer returned by the function, check for NULL and if so return None, otherwise the `&*` + cast

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -165,167 +164,167 @@ impl SDPMessage {
-    pub fn get_information(&self) -> Option<String> {
+    pub fn get_information(&self) -> Option<&str> {
         unsafe {
             from_glib_none(ffi::gst_sdp_message_get_information(self.to_glib_none().0))
```
And for strings, check for NULL first and then use `CStr` directly

notbdu commented:
Thanks for the additional comments, they're very helpful! I'm still working through all the issues.... there's quite a few haha 😓 .

sdroege commented:
Don't worry and take your time, and ask if you have any questions :)

notbdu commented:
Alright, so I got it down to just 4 compilation errors now...
```
error[E0599]: no method named `to_glib_none_mut` found for type `&mut gst::CapsRef` in the current scope
   --> gstreamer-sdp/src/s_d_p_message.rs:127:91
    |
127 |             from_glib(ffi::gst_sdp_message_attributes_to_caps(self.to_glib_none().0, caps.to_glib_none_mut().0))
    |                                                                                           ^^^^^^^^^^^^^^^^

error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> gstreamer-sdp/src/s_d_p_media.rs:31:23
   |
31 |         copy => |ptr| ffi::gst_sdp_media_copy(mut_override(ptr)),
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 parameters

error[E0308]: mismatched types
  --> gstreamer-sdp/src/s_d_p_media.rs:31:23
   |
27 | / glib_wrapper! {
28 | |     pub struct SDPMedia(Boxed<ffi::GstSDPMedia>);
29 | |
30 | |     match fn {
31 | |         copy => |ptr| ffi::gst_sdp_media_copy(mut_override(ptr)),
   | |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected *-ptr, found i32
32 | |         free => |ptr| ffi::gst_sdp_media_free(ptr),
33 | |     }
34 | | }
   | |_- expected `*mut ffi::GstSDPMedia` because of return type
   |
   = note: expected type `*mut ffi::GstSDPMedia`
              found type `i32`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0308]: mismatched types
  --> gstreamer-sdp/src/s_d_p_media.rs:32:23
   |
27 | / glib_wrapper! {
28 | |     pub struct SDPMedia(Boxed<ffi::GstSDPMedia>);
29 | |
30 | |     match fn {
31 | |         copy => |ptr| ffi::gst_sdp_media_copy(mut_override(ptr)),
32 | |         free => |ptr| ffi::gst_sdp_media_free(ptr),
   | |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^- help: try adding a semicolon: `;`
   | |                       |
   | |                       expected (), found i32
33 | |     }
34 | | }
   | |_- expected `()` because of default return type
   |
   = note: expected type `()`
              found type `i32`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 4 previous errors
```

notbdu commented:
A few questions:
- Should I add the `ToGlibPtr` trait impl to `CapsRef` or should I just expose the wrapped `ffi::GstCapsRef` struct?
- I ended up exposing all of the wrapped `ffi` structs to avoid having to impl all of the to glib traits... is that okay for now? I'm assuming we'd want to use the `glib_wrapper` macro on everything?
- It looks like the auto generated copy/free functions for `SDPMedia`are incorrect. I'm not sure what to do there...

sdroege commented:
```
error[E0599]: no method named `to_glib_none_mut` found for type `&mut gst::CapsRef` in the current scope
   --> gstreamer-sdp/src/s_d_p_message.rs:127:91
    |
127 |             from_glib(ffi::gst_sdp_message_attributes_to_caps(self.to_glib_none().0, caps.to_glib_none_mut().0))
    |     
```

Use `caps.as_mut_ptr()` here

```
error[E0308]: mismatched types
  --> gstreamer-sdp/src/s_d_p_media.rs:31:23
   |
27 | / glib_wrapper! {
28 | |     pub struct SDPMedia(Boxed<ffi::GstSDPMedia>);
29 | |
30 | |     match fn {
31 | |         copy => |ptr| ffi::gst_sdp_media_copy(mut_override(ptr)),
   | |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected *-ptr, found i32
32 | |         free => |ptr| ffi::gst_sdp_media_free(ptr),
33 | |     }
34 | | }
   | |_- expected `*mut ffi::GstSDPMedia` because of return type
   |
   = note: expected type `*mut ffi::GstSDPMedia`
              found type `i32`
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

The signature of the functions is wrong. You could put a closure there that, that in the case of free ignores the return value and in case of copy first mallocs memory, then passes it there, and then returns it while ignoring the return value.

sdroege commented:
For your questions, I think all but this one are answered by my comment?

> I ended up exposing all of the wrapped ffi structs to avoid having to impl all of the to glib traits... is that okay for now? I'm assuming we'd want to use the glib_wrapper macro on everything?

We can't use `glib_wrapper` for most of these types and simply wrapping the ffi types is the only thing that can be done here. That's all ok :)

notbdu commented:
Okay, it compiles now w/ some warnings (mostly just unused code)...

sdroege commented:
Thanks a lot, especially for your patience. I'll go through everything tomorrow. Before merging, we should get rid of the warnings though but that should be easy :)

notbdu commented:
No problem, I should be thanking you for reviewing my PR and helping me work through the issues.

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_payload_s_p_param.rs:
```diff
@@ -12,7 +12,7 @@ use ffi;
 
 impl MIKEYPayloadSPParam {
-    pub fn new(type_: u8, len: u8, val: u8) -> MIKEYPayloadSPParam {
+    pub fn new(type_: u8, len: u8, mut val: u8) -> MIKEYPayloadSPParam {
```
This is not a simple `u8` but a `&[u8]`. I don't think we need a constructor at all for this as `gst_mikey_payload_sp_add_param()` internally creates it anyway.

sdroege commented on gstreamer-sdp/src/m_i_k_e_y_payload_s_p_param.rs:
```diff
@@ -29,6 +29,8 @@ impl MIKEYPayloadSPParam {
     pub fn val(&self) -> u8 {
-        *self.0.val
+        unsafe {
+            *self.0.val
```
This should use `slice::from_raw_parts` and return a `&[u8]`

notbdu replied:
Ah, dammit I still have to get used to the idea of a u8 pointer actually being an array/slice instead of a single u8... t.t

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -28,8 +28,14 @@ glib_wrapper! {
+        copy => |ptr| {
+            let mut copy = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMedia>()) as *mut ffi::GstSDPMedia;
+            ffi::gst_sdp_media_copy(ptr as *const ffi::GstSDPMedia, &mut copy);
+            copy
```
This is actually wrong, sorry.
```rust
            let mut copy = ptr::null_mut();
            ffi::gst_sdp_media_copy(ptr as *const ffi::GstSDPMedia, &mut copy);
            copy
```

notbdu replied:
Curious, when is it idiomatic to use the following different methods for passing memory to a `ffi` "constructor" method. After observing the usage of each I haven't made any sensible conclusions so far... XD

  - `mem::uninitialized()`
  - `mem::zeroed()` (It looks like this method is always preferred over `mem::uninitialized`? It's possible for uninitialized memory to contain garbage?)
  - `ptr::null_mut()`
  - `g_malloc0` (After reading docs [here](https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html#g-malloc) I can see why this is preferred)

sdroege replied:
`uninitialized`/`g_malloc` are fine if the memory is going to be fully initialized. `zeroed`/`g_malloc0` otherwise. The `mem::` functions for stack allocations, `g_malloc` for heap allocation.

sdroege replied:
Oh and I don't really trust the SDP library to properly initialize everything. It seems to require you to zero-initialize things. As said, uncommon and annoying API :)

notbdu replied:
Hmmmm, what is the rationale for using `ptr::null_mut()` here versus `g_malloc`. Both methods are allocating memory on the heap?

sdroege replied:
First is on the stack, second on the heap. The first here because the function will allocate for us and we only need to provide a pointer where it can write to

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -465,7 +471,7 @@ impl SDPMedia {
     }
 
-    pub fn set_media_from_caps(caps: &gst::Caps, media: SDPMedia) -> Result<(), ()> {
+    pub fn set_media_from_caps(caps: &gst::Caps, mut media: SDPMedia) -> Result<(), ()> {
```
This should take the media by reference. We fill the existing media with information from the caps.

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -667,7 +668,7 @@ impl SDPMessage {
         unsafe {
             let size = data.len() as u32;
-            let mut msg = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMessage>()) as *mut ffi::GstSDPMessage;
+            let msg = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMessage>()) as *mut ffi::GstSDPMessage;
```
This should use `g_malloc0`, also in the other places

sdroege replied:
I guess most of the `mem::uninitialized` here should be `mem::zeroed` instead

notbdu commented:
Gotcha, that makes sense to me. Thanks!!

On Wed, Apr 4, 2018 at 12:36 PM, Sebastian Dröge <notifications@github.com>
wrote:

> *@sdroege* commented on this pull request.
> ------------------------------
>
> In gstreamer-sdp/src/s_d_p_media.rs
> <https://github.com/sdroege/gstreamer-rs/pull/100#discussion_r179206149>:
>
> > @@ -28,8 +28,14 @@ glib_wrapper! {
>      pub struct SDPMedia(Boxed<ffi::GstSDPMedia>);
>
>      match fn {
> -        copy => |ptr| ffi::gst_sdp_media_copy(mut_override(ptr)),
> -        free => |ptr| ffi::gst_sdp_media_free(ptr),
> +        copy => |ptr| {
> +            let mut copy = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMedia>()) as *mut ffi::GstSDPMedia;
> +            ffi::gst_sdp_media_copy(ptr as *const ffi::GstSDPMedia, &mut copy);
> +            copy
>
> Oh and I don't really trust the SDP library to properly initialize
> everything. It seems to require you to zero-initialize things. As said,
> uncommon and annoying API :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sdroege/gstreamer-rs/pull/100#discussion_r179206149>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AILs7TjlxEQrI9Ohz4tN7k3NOl3YJ7GFks5tlPaHgaJpZM4S6Mx_>
> .
>

sdroege commented on gstreamer-sdp/src/s_d_p_connection.rs:
```diff
@@ -10,17 +10,18 @@ use std::mem;
-
-pub struct SDPConnection(ffi::GstSDPConnection);
+#[repr(C)]
+pub struct SDPConnection(pub ffi::GstSDPConnection);
```
IMHO these wrapper structs should not make their content `pub`. If anything, make them `pub(crate)` as needed

notbdu replied:
Oh nice! Didn't know about `pub(crate)`.

sdroege commented on gstreamer-sdp/src/s_d_p_media.rs:
```diff
@@ -25,55 +28,66 @@ glib_wrapper! {
+        unsafe {
+            let mut media = ptr::null_mut();
+            let ret = from_glib(ffi::gst_sdp_media_new(&mut media));
+            (ret, from_glib_full(media))
```
This should return a `Result`, not a tuple

sdroege commented on gstreamer-sdp/src/s_d_p_message.rs:
```diff
@@ -588,27 +664,27 @@ impl SDPMessage {
-                SDPResult::ok => Some(from_glib_full(msg)),
-                _ => Err(result),
+                SDPResult::Ok => Ok(from_glib_full(msg)),
+                _ => Err(()),
```
On error you need to `glib_ffi::g_free` the pointer, everywhere where `g_malloc` is used

notbdu replied:
That is true...

sdroege commented:
If you rebase on the `webrtc-sdp` branch now (and make sure that only my `WebRTC/SDP WIP` commit stays, none of the others from me), it should even build on travis.

sdroege commented:
This still does not build. Did you address all the review comments though?

sdroege commented:
You did, thanks :) Looks mostly good to me and I'll fix the remaining build problems and do some minor cleanup. The build problems are my fault actually

sdroege commented:
And all merged, thanks a lot :) I did some minor cleanup, and will do a second go over the API again tomorrow.
